astropy/nddata/covariance.py

```diff
@@ -0,0 +1,1034 @@
+            in a two-dimensional image. The shape of the image is expected to be
+            square, such that the shape of the covariance matrix is
+            :math:`N_x\times N_y`, where :math:`N_x = N_y`. Each of the ``I``
+            and ``J`` output columns are then split into two columns according
+            to associate coordinates, such that there are 8 output columns.
+
+        hdr : `astropy.io.fits.Header`, optional
+            `astropy.io.fits.Header` instance to which to add covariance
+            keywords. If None, a new `astropy.io.fits.Header` instance is
+            returned.
+
+        Returns
+        -------
+        tuple
+            Returns three objects: (1) The header for the primary HDU; (2) an
+            `astropy.io.fits.ImageHDU` object with the variance vector; and (3)
+            an `astropy.io.fits.BinTableHDU` object with the correlation
+            coefficients.
+        """
+        # Use input header or create a minimal one
+        _hdr = fits.Header() if hdr is None else hdr
+        # Ensure the input header has the correct type
+        if not isinstance(_hdr, fits.Header):
+            raise TypeError("Input header must have type astropy.io.fits.Header.")
+        _hdr["COVSHAPE"] = (str(self.shape), "Shape of the correlation matrix")
+
+        tbl_hdr = fits.Header()
+        tbl_hdr["COVSHAPE"] = (str(self.shape), "Shape of the correlation matrix")
+        if self.raw_shape is not None:
+            tbl_hdr["COVRWSHP"] = (str(self.raw_shape), "Raw shape of the source data")
+
+        # Construct the correlation binary table HDU
+        if reshape:
+            i_c1, i_c2, j_c1, j_c2, rhoij, var = self.coordinate_data(reshape=reshape)
+            covar_hdu = fits.BinTableHDU.from_columns(
+                [
+                    fits.Column(name="INDXI_C1", format="1J", array=i_c1),
+                    fits.Column(name="INDXI_C2", format="1J", array=i_c2),
+                    fits.Column(name="INDXJ_C1", format="1J", array=j_c1),
+                    fits.Column(name="INDXJ_C2", format="1J", array=j_c2),
+                    fits.Column(name="RHOIJ", format="1D", array=rhoij),
+                ],
+                name="CORREL",
+                header=tbl_hdr,
+            )
+        else:
+            i, j, rhoij, var = self.coordinate_data(reshape=reshape)
+            covar_hdu = fits.BinTableHDU.from_columns(
+                [
+                    fits.Column(name="INDXI", format="1J", array=i),
+                    fits.Column(name="INDXJ", format="1J", array=j),
+                    fits.Column(name="RHOIJ", format="1D", array=rhoij),
+                ],
+                name="CORREL",
+                header=tbl_hdr,
+            )
+
+        ivar_hdu = fits.ImageHDU(data=np.ma.power(var, -1.0).filled(0.0), name="IVAR")
+        return _hdr, ivar_hdu, covar_hdu
+
+    def write(self, ofile, reshape=False, hdr=None, overwrite=False):
```

Comment:
This will perhaps need most discussion – I think `NDData` objects, or rather their subclasses, are generally (expected to be) using the unified IO framework, and having a class bring in its own `write` method will probably not fit in with that. More specifically, #16648 envisions `Covariance` to be associated with an `NDData` object, to be read and written together.
I _think_ this means such subclassed objects would set up their own readers and writers and register them, as it is done e.g. for `CCDData` and would be done for `Spectrum1D`, and handle themselves exactly how and where to store a `covariance` attribute, just as it is done for `uncertainty`. The `output_hdus` method defined above would be crucial for that, as it is arguably a bit more complex than just storing an uncertainty array as table row or data field.

Reply:
I restructured this a bit.  For how I'm planning to use this in `specutils`, I just needed a to/from tables routines.  I've now implemented these, and the write/from_fits functions largely just wrap these functions.  If these FITS I/O functions are an issue, we can just remove them.

The bigger issue may be that I don't know how NDData interfaces with the unified I/O framework.  I'm now subclassing `Covariance` from `NDUncertainty`, but this may play havoc with `NDData` I/O if the `uncertainty` attribute is a `Covariance` instance.